code/controllers/subsystem/movement/zfall.dm

```diff
@@ -0,0 +1,37 @@
+/// Responsible for triggering mob zfalls when they are not already triggered by structure deconstruction or moving into openspace
+SUBSYSTEM_DEF(zfall)
+	name = "Zfall"
+	flags = SS_TICKER
+	wait = 5
+	runlevels = RUNLEVEL_GAME|RUNLEVEL_POSTGAME
+	priority = FIRE_PRIORITY_ZFALL
+	var/list/atom/movable/openspace_inhabitants = list()
+
+/datum/controller/subsystem/zfall/proc/add_openspace_inhabitant(atom/movable/A)
+	if(!ismob(A) || (A in openspace_inhabitants))
+		return
+	RegisterSignal(A, COMSIG_PARENT_QDELETING, .proc/remove_openspace_inhabitant, A)
+	openspace_inhabitants |= A
+
+/datum/controller/subsystem/zfall/proc/remove_openspace_inhabitant(atom/movable/A)
+	SIGNAL_HANDLER
+	if(!(A in openspace_inhabitants))
+		return
+	UnregisterSignal(A, COMSIG_PARENT_QDELETING)
+	openspace_inhabitants -= A
+
+/datum/controller/subsystem/zfall/fire(resumed)
+	var/list/removals = list()
+	for(var/atom/movable/A as() in openspace_inhabitants)
+		var/turf/T = get_turf(A)
+		if(!isopenspace(T))
+			removals |= A
+			continue
+		if(!A.zfalling)
+			T.zFall(A)
+		if(MC_TICK_CHECK)
+			return
+	for(var/A as() in removals)
+		remove_openspace_inhabitant(A)
+		if(MC_TICK_CHECK)
+			return
```

Comment:
Should this whole thing not just be signal based? If something is destroyed then we can react to that rather than checking every tick if the thing below us still exists?

Reply:
Every one of those cases is capable of sending a signal that can be reacted to

Reply:
Movement type changes are not a signal, as is not jetpack toggling, jetpack fuel loss,  cyborg ion thruster disabling, etc

It's minimal performance cost for 100x easier maintainability. The max amount of mobs occupying openspace at a given time will be like 4

Reply:
Also jetpacks need to use fuel as long as they are in openspace and this subsystem does that timing

Reply:
imo if you are registering >5 signals on one object you are doing it wrong. this would require an absolutely astronomical amount of signals, the signal system is not really that robust. for example unregistering a signal does not guarantee the one YOU registered will be unregistered because it doesn't have IDs

Reply:
It does since it associates with the datum you registered from.
If you have 2 different datums that register for the same signal, then calling RemoveComponent on one of the datums will remove the correct signal, not both of them.

Reply:
Worked out that we'll just do an enumeration subsystem